courses/fundamentals_of_ada/240b_tasking.rst

```diff
@@ -0,0 +1,492 @@
+     
+* May be executed as follows:
+    
+   .. code:: Ada
+    
+      calling start
+      start             -- May switch place with line below
+      calling receive 1 -- May switch place with line above
+      Receive 1
+      calling receive 2
+      -- Blocked until another task calls Start
+     
+------------------------
+Accepting a Rendezvous
+------------------------
+
+* :ada:`accept` statement
+
+   - Wait on single entry
+   - If entry call waiting: Server handles it
+   - Else: Server **waits** for an entry call
+
+* :ada:`select` statement
+
+   - **Several** entries accepted at the **same time**
+   - Can **time-out** on the wait
+   - Can be **not blocking** if no entry call waiting
+   - Can **terminate** if no clients can **possibly** make entry call
+   - Can **conditionally** accept a rendezvous based on a **guard expression**
+
+===================
+Protected Objects
+===================
+
+-------------------
+Protected Objects
+-------------------
+
+* **Passive** objects state
+
+   - **Multitask-safe** accessors to get and set state
+   - **No** direct state manipulation
+   - **No** concurrent modifications
+
+* Protected objects are :ada:`limited` types
+
+.. code:: Ada
+    
+   protected Object is
+      procedure Set (V : Integer);
+      function Get return Integer;
+   private
+      Value : Integer;
+   end Object;
+       
+   protected body Object is
+      procedure Set (V : Integer) is
+      begin
+         Value := V;
+      end Set;
+       
+      function Get return Integer is
+      begin
+         return Value;
+      end Get;
+   end Object;
+     
+-------------------------------------
+Protected: Functions and Procedures
+-------------------------------------
+
+* A :ada:`function` can **get** the state
+
+   - Protected data is **read-only**
+   - Concurrent call to :ada:`function` is **allowed**
+   - **No** concurrent call to :ada:`procedure`
+
+* A :ada:`procedure` can **set** the state
+
+   - **No** concurrent call to either :ada:`procedure` or :ada:`function`
+
+* In case of concurrency, other callers get **blocked**
+
+    - Until call finishes
+
+======
+Delays
+======
+
+-------------
+Delay keyword
+-------------
+
+- :ada:`delay` keyword part of tasking
+- Blocks for a time
+- Relative: Blocks for a :ada:`Duration`
+- Absolute: Blocks until a given :ada:`Calendar.Time` or :ada:`Real_Time.Time`
+
+.. code:: Ada
+
+    Relative : Duration := Seconds(5.0);
+    delay Relative;
+
+    Absolute : Time := Time_Of (2030, 10, 30);
+    delay until Absolute;
+
+
+==========================
+Task and Protected Types
+==========================
+
+---------------
+Task Activation
+---------------
+
+* An instanciated task starts running when **activated**
+* On the stack
+
+    - Activated when **enclosing** declarative part finishes its **elaboration**
+
+* On the heap
+
+    - Activated **immediately** at instanciation
+
+.. code:: Ada
+    
+   task type T is [...]
```

Comment:
We haven't talked about tasks vs task types yet. To keep this "basic", update the previous examples to always use "task type T"

Reply:
Ok fair enough, and I'll add some note about `task T` implying the declaration of an anonymous task type with a named instance T.